FreeRTOS-Plus/Demo/AWS/Device_Defender_Windows_Simulator/Device_Defender_Demo/report_builder.c

```diff
@@ -484,20 +622,65 @@ eReportBuilderStatus eGenerateJsonReport( char * pcBuffer,
     /* Write part4. */
     if( eStatus == eReportBuilderSuccess )
     {
-        ulCharactersWritten = snprintf( pcCurrentWritePos,
+        lCharactersWritten = snprintf( pcCurrentWritePos,
+                                       ulRemainingBufferLength,
+                                       reportbuilderJSON_REPORT_FORMAT_PART4,
+                                       DEFENDER_REPORT_TOTAL_KEY,
+                                       pxMetrics->ulEstablishedConnectionsArrayLength,
+                                       DEFENDER_REPORT_CUSTOM_METRICS_KEY,
+                                       DEFENDER_REPORT_NUMBER_KEY,
+                                       pxMetrics->ulStackHighWaterMark,
+                                       DEFENDER_REPORT_NUMBER_LIST_KEY
+                                       );
+
+        if( !reportbuilderSNPRINTF_SUCCESS( lCharactersWritten, ulRemainingBufferLength ) )
+        {
+            LogError( ( "Failed to write part 4." ) );
+            eStatus = eReportBuilderBufferTooSmall;
+        }
+        else
+        {
+            ulRemainingBufferLength -= lCharactersWritten;
+            pcCurrentWritePos += lCharactersWritten;
+        }
+    }
+
+    /* Write task ids array. */
+    if( eStatus == eReportBuilderSuccess )
+    {
+        eStatus = prvWriteTaskIdsArray( pcCurrentWritePos,
                                         ulRemainingBufferLength,
-                                        reportbuilderJSON_REPORT_FORMAT_PART4,
-                                        pxMetrics->ulEstablishedConnectionsArrayLength );
+                                        pxMetrics->pulTaskIdsArray,
+                                        pxMetrics->ulTaskIdsArrayLength,
+                                        &( bufferWritten ) );
 
-        if( !reportbuilderSNPRINTF_SUCCESS( ulCharactersWritten, ulRemainingBufferLength ) )
+        if( eStatus == eReportBuilderSuccess )
         {
-            LogError( ( "Failed to write part 4." ) );
+            pcCurrentWritePos += bufferWritten;
+            ulRemainingBufferLength -= bufferWritten;
+        }
+        else
+        {
+            LogError( ( "Failed to write task ids array." ) );
+        }
+    }
+
+    /* Write part5. */
+    if( eStatus == eReportBuilderSuccess )
+    {
+        lCharactersWritten = snprintf( pcCurrentWritePos,
+                                       ulRemainingBufferLength,
+                                       reportbuilderJSON_REPORT_FORMAT_PART5 );
+
+        if( !reportbuilderSNPRINTF_SUCCESS( lCharactersWritten, ulRemainingBufferLength ) )
+        {
+            LogError( ( "Failed to write part 5." ) );
             eStatus = eReportBuilderBufferTooSmall;
         }
         else
         {
-            ulRemainingBufferLength -= ulCharactersWritten;
-            pcCurrentWritePos += ulCharactersWritten;
+            ulRemainingBufferLength -= lCharactersWritten;
+            pcCurrentWritePos += lCharactersWritten;
```

Comment:
You can simplify this logic by moving line 689 between lines 683/684 and deleting lines 687-688 & 690.